File: scripts/hent_klinikker.js
/**
 * Henter klinikker fra Brønnøysundregistrene og oppdaterer HTML-filene.
 * Bruk: node scripts/hent_klinikker.js --kommunenavn drammen
 */

const https = require("https");
const fs = require("fs");
const path = require("path");

const NACE_KODER = ["96.220", "96.230"];

// Poststeder brukt som filter mot Brønnøysund API
const POSTSTEDER = {
  drammen:   "DRAMMEN",
  oslo:      "OSLO",
  bergen:    "BERGEN",
  trondheim: "TRONDHEIM",
  stavanger: "STAVANGER",
};

const BASE_URL = "https://data.brreg.no/enhetsregisteret/api/enheter";

function hentJSON(url) {
  return new Promise((resolve, reject) => {
    https.get(url, { headers: { Accept: "application/json" } }, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => {
        try { resolve(JSON.parse(data)); }
        catch (e) { reject(e); }
      });
    }).on("error", reject);
  });
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function hentAlleForNace(naceKode, poststed) {
  const resultater = [];
  let side = 0;
  process.stdout.write(`  Henter NACE ${naceKode} ...`);

  while (true) {
    const params = new URLSearchParams({
      naeringskode: naceKode,
      "forretningsadresse.poststed": poststed,
      size: 100,
      page: side,
    });
    const url = `${BASE_URL}?${params}`;
    const data = await hentJSON(url);
    const enheter = data?._embedded?.enheter ?? [];
    const totalSider = data?.page?.totalPages ?? 0;

    for (const enhet of enheter) {
      if (enhet.konkurs || enhet.underAvvikling) continue;
      resultater.push(enhet);
    }

    side++;
    if (side >= totalSider) break;
    await sleep(200);
  }

  console.log(` ${resultater.length} treff`);
  return resultater;
}

function lagInitialer(navn) {
  const ord = navn.replace(/"/g, "").split(/\s+/);
  return ord.map((o) => o[0]).filter(Boolean).join("").slice(0, 2).toUpperCase() || "??";
}

function formatAdresse(adr) {
  if (!adr) return "Ikke oppgitt";
  const gate = (adr.adresse ?? []).join(", ");
  return [gate, adr.postnummer, adr.poststed].filter(Boolean).join(" ").trim() || "Ikke oppgitt";
}

function finnTjenester(enhet) {
  const nace = enhet.naeringskode1?.kode ?? "";
  const base = {
    "96.220": ["Hudpleie", "Ansiktsbehandling"],
    "96.230": ["Spa", "Massasje"],
    "86.210": ["Medisinsk estetikk", "Botox"],
  }[nace] ?? ["Skjønnhetspleie"];

  const extra = (enhet.aktivitet ?? [])
    .slice(0, 2)
    .map((a) => a.replace(/\.$/, "").trim())
    .filter((a) => a && !base.includes(a));

  return [...base, ...extra].slice(0, 4);
}

function genererKlinikkort(enhet, verifisert = false) {
  const navn = enhet.navn ?? "Ukjent";
  const orgnrRaa = enhet.organisasjonsnummer ?? "";
  const orgnr = orgnrRaa.length === 9
    ? `${orgnrRaa.slice(0, 3)} ${orgnrRaa.slice(3, 6)} ${orgnrRaa.slice(6)}`
    : orgnrRaa;
  const initialer = lagInitialer(navn);
  const adresse = formatAdresse(enhet.forretningsadresse);
  const telefon = enhet.telefon || enhet.mobil || enhet.verifisering?.telefon_funnet || "Ikke oppgitt";
  const nettside = enhet.nettside_funnet || enhet.hjemmeside || "Ikke oppgitt";
  const tjenester = finnTjenester(enhet);
  const navnTittel = navn.toLowerCase().replace(/\b\w/g, (c) => c.toUpperCase());
  const badge = verifisert ? "✓ Verifisert" : "✓ Registrert";
  const badgeKlasse = verifisert ? "verified-badge verified-badge--ok" : "verified-badge";

  const tjenesteTags = tjenester
    .map((t) => `            <span class="service-tag">${t}</span>`)
    .join("\n");

  return `        <div class="clinic-card">
          <div class="card-header">
            <div class="clinic-initials">${initialer}</div>
            <span class="${badgeKlasse}">${badge}</span>
          </div>
          <div class="clinic-name">${navnTittel}</div>
          <div class="clinic-orgnr">Org.nr: ${orgnr}</div>
          <div class="card-details">
            <div class="detail-row">📍 <span>${adresse}</span></div>
            <div class="detail-row">📞 <span>${telefon}</span></div>
            <div class="detail-row">🌐 <span>${nettside}</span></div>
          </div>
          <div class="services-row">
${tjenesteTags}
          </div>
          <div class="card-footer">
            <button class="btn-primary">Se detaljer</button>
            <button class="btn-secondary">Kart</button>
          </div>
        </div>`;
}

// Konfigurasjon per sidetype: hvilke NACE-koder og nøkkelord som er relevante
const SITE_CONFIG = {
  botox:    { nace: ["96.220", "96.230"], nøkkelord: "botox"    },
  laser:    { nace: ["96.220", "96.230"], nøkkelord: "laser"    },
  hudpleie: { nace: ["96.220", "96.230"], nøkkelord: "hudpleie" },
};

async function oppdaterFraEnriched(kommunenavn) {
  const dataMappe = path.join(__dirname, "data");
  const enrichedFil = path.join(dataMappe, `klinikker_${kommunenavn}_enriched.json`);

  if (!fs.existsSync(enrichedFil)) {
    console.error(`Finner ikke ${enrichedFil} – kjør finn_nettsider.js og verifiser_nettsider.js først`);
    process.exit(1);
  }

  const data = JSON.parse(fs.readFileSync(enrichedFil, "utf-8"));
  const rot = path.resolve(__dirname, "..");

  for (const mappe of fs.readdirSync(rot)) {
    if (mappe.startsWith("_") || mappe.startsWith(".")) continue;
    if (!mappe.toLowerCase().includes(kommunenavn.toLowerCase())) continue;

    const mappeLower = mappe.toLowerCase();
    let sidetype = null;
    for (const type of Object.keys(SITE_CONFIG)) {
      if (mappeLower.startsWith(type)) { sidetype = type; break; }
    }
    if (!sidetype) continue;

    const config = SITE_CONFIG[sidetype];
    const filtrert = data.filter(
      (e) =>
        config.nace.includes(e.naeringskode1?.kode) &&
        e.verifisering?.nøkkelord_treff?.[config.nøkkelord] === true
    );
    filtrert.sort((a, b) => a.navn.localeCompare(b.navn));

    const kort = filtrert.map((e) => genererKlinikkort(e, true)).join("\n");
    const htmlFil = path.join(rot, mappe, "index.html");
    if (fs.existsSync(htmlFil)) {
      oppdaterHTML(htmlFil, kort, filtrert.length);
      console.log(`  ${mappe}: ${filtrert.length} verifiserte klinikker`);
    }
  }
}

function oppdaterHTML(htmlFil, klinikkortHTML, antall) {
  let innhold = fs.readFileSync(htmlFil, "utf-8");

  // Erstatt klinikkgrid – bruker unik avslutnings-kontekst som anker
  innhold = innhold.replace(
    /(<div class="clinic-grid">)[\s\S]*?(    <\/div>\n  <\/div>\n<\/section>)/,
    `$1\n\n${klinikkortHTML}\n\n    </div>\n  </div>\n</section>`
  );

  // Oppdater antall i hero-statistikk: <div class="stat-num">12</div>
  innhold = innhold.replace(
    /(<div class="stat-num">)\d+(<\/div>)/,
    `$1${antall}$2`
  );

  // Oppdater result-count
  innhold = innhold.replace(
    /(<span class="result-count">Viser )\d+( klinikker<\/span>)/,
    `$1${antall}$2`
  );

  fs.writeFileSync(htmlFil, innhold, "utf-8");
  console.log(`  Oppdatert: ${htmlFil}`);
}

async function main() {
  const args = {};
  const argv = process.argv.slice(2);
  for (let i = 0; i < argv.length; i++) {
    if (argv[i].startsWith("--")) {
      const key = argv[i].slice(2);
      const next = argv[i + 1];
      args[key] = next && !next.startsWith("--") ? (i++, next) : true;
    }
  }

  const kommunenavn = (args.kommunenavn ?? args.by ?? "").toLowerCase();
  let poststed = args.poststed ?? POSTSTEDER[kommunenavn];

  if (!kommunenavn) {
    console.log("Bruk: node scripts/hent_klinikker.js --kommunenavn drammen [--fra-enriched]");
    console.log(`Kjente byer: ${Object.keys(POSTSTEDER).join(", ")}`);
    process.exit(1);
  }

  if ("fra-enriched" in args) {
    console.log(`\nOppdaterer HTML fra enriched data for ${kommunenavn}...`);
    await oppdaterFraEnriched(kommunenavn);
    console.log("\nFerdig!");
    return;
  }

  if (!poststed) poststed = kommunenavn.toUpperCase();

  console.log(`\nHenter klinikker for ${poststed}...`);

  const alle = [];
  for (const nace of NACE_KODER) {
    const treff = await hentAlleForNace(nace, poststed);
    alle.push(...treff);
  }

  // Dedupliser på org.nr
  const sett = new Map();
  for (const e of alle) sett.set(e.organisasjonsnummer, e);
  const unike = [...sett.values()].sort((a, b) => a.navn.localeCompare(b.navn));

  console.log(`\nTotalt ${unike.length} unike klinikker funnet.`);

  // Lagre rådata
  const rot = path.resolve(__dirname, "..");
  const dataMappe = path.join(__dirname, "data");
  if (!fs.existsSync(dataMappe)) fs.mkdirSync(dataMappe, { recursive: true });

  const jsonFil = path.join(dataMappe, `klinikker_${kommunenavn || poststed.toLowerCase()}.json`);
  fs.writeFileSync(jsonFil, JSON.stringify(unike, null, 2), "utf-8");
  console.log(`  Rådata lagret: ${jsonFil}`);

  // Generer HTML-klinikkort
  const alleKort = unike.map(genererKlinikkort).join("\n");

  // Finn og oppdater relevante HTML-filer
  for (const mappe of fs.readdirSync(rot)) {
    if (mappe.startsWith("_") || mappe.startsWith(".")) continue;
    const mappeNavnLower = mappe.toLowerCase();
    if (!kommunenavn || !mappeNavnLower.includes(kommunenavn.toLowerCase())) continue;
    const htmlFil = path.join(rot, mappe, "index.html");
    if (fs.existsSync(htmlFil)) {
      oppdaterHTML(htmlFil, alleKort, unike.length);
    }
  }

  console.log("\nFerdig!");
}

main().catch((err) => {
  console.error("Feil:", err.message);
  process.exit(1);
});

File: scripts/hent_klinikker.py
"""
Henter klinikker fra Brønnøysundregistrene og genererer HTML-klinikkort.
Bruk: python scripts/hent_klinikker.py --kommune 3005 --kommunenavn Drammen
"""

import argparse
import json
import time
import urllib.request
import urllib.parse
from pathlib import Path

NACE_KODER = {
    "96.220": "Skjønnhetspleie",
    "96.230": "Spa og badstue",
    "86.210": "Legevirksomhet (estetikk)",
}

KOMMUNER = {
    "drammen":    "3005",
    "oslo":       "0301",
    "bergen":     "4601",
    "trondheim":  "5001",
    "stavanger":  "1103",
}

BASE_URL = "https://data.brreg.no/enhetsregisteret/api/enheter"


def hent_side(nace_kode: str, side: int, stoerrelse: int = 100) -> dict:
    params = urllib.parse.urlencode({
        "naeringskode": nace_kode,
        "size": stoerrelse,
        "page": side,
    })
    url = f"{BASE_URL}?{params}"
    req = urllib.request.Request(url, headers={"Accept": "application/json"})
    with urllib.request.urlopen(req, timeout=15) as resp:
        return json.loads(resp.read().decode())


def hent_alle_for_nace(nace_kode: str, kommunenummer: str) -> list[dict]:
    """Henter alle aktive bedrifter for en NACE-kode og filtrerer på kommunenummer."""
    resultater = []
    side = 0

    print(f"  Henter NACE {nace_kode} ...", end="", flush=True)
    while True:
        data = hent_side(nace_kode, side)
        enheter = data.get("_embedded", {}).get("enheter", [])
        total_sider = data.get("page", {}).get("totalPages", 0)

        for enhet in enheter:
            if enhet.get("konkurs") or enhet.get("underAvvikling"):
                continue
            adresse = enhet.get("forretningsadresse", {})
            if adresse.get("kommunenummer") == kommunenummer:
                resultater.append(enhet)

        side += 1
        if side >= total_sider:
            break
        time.sleep(0.2)  # Ikke overbelast APIet

    print(f" {len(resultater)} treff")
    return resultater


def lag_initialer(navn: str) -> str:
    ord_liste = navn.replace('"', "").split()
    initialer = "".join(o[0] for o in ord_liste if o[0].isalpha())
    return initialer[:2].upper() if initialer else "??"


def format_adresse(adresse: dict) -> str:
    gate = ", ".join(adresse.get("adresse", []))
    postnr = adresse.get("postnummer", "")
    poststed = adresse.get("poststed", "")
    if gate:
        return f"{gate}, {postnr} {poststed}".strip(", ")
    return f"{postnr} {poststed}".strip()


def finn_tjenester(enhet: dict) -> list[str]:
    nace = enhet.get("naeringskode1", {}).get("kode", "")
    tjenester = list({
        "96.220": ["Hudpleie", "Ansiktsbehandling", "Kroppsbehandling"],
        "96.230": ["Spa", "Massasje", "Velvære"],
        "86.210": ["Medisinsk estetikk", "Botox", "Fillers"],
    }.get(nace, ["Skjønnhetspleie"]))
    aktiviteter = enhet.get("aktivitet", [])
    if aktiviteter:
        for akt in aktiviteter[:2]:
            kort = akt.rstrip(".").strip()
            if kort and kort not in tjenester:
                tjenester.append(kort)
    return tjenester[:4]


def generer_klinikkort(enhet: dict) -> str:
    navn = enhet.get("navn", "Ukjent")
    orgnr_raa = enhet.get("organisasjonsnummer", "")
    orgnr = f"{orgnr_raa[:3]} {orgnr_raa[3:6]} {orgnr_raa[6:]}" if len(orgnr_raa) == 9 else orgnr_raa
    initialer = lag_initialer(navn)
    adresse = format_adresse(enhet.get("forretningsadresse", {}))
    telefon = enhet.get("telefon", "") or enhet.get("mobil", "") or "Ikke oppgitt"
    nettside = enhet.get("hjemmeside", "") or "Ikke oppgitt"
    tjenester = finn_tjenester(enhet)

    tjeneste_tags = "\n            ".join(
        f'<span class="service-tag">{t}</span>' for t in tjenester
    )

    return f"""        <div class="clinic-card">
          <div class="card-header">
            <div class="clinic-initials">{initialer}</div>
            <span class="verified-badge">✓ Registrert</span>
          </div>
          <div class="clinic-name">{navn.title()}</div>
          <div class="clinic-orgnr">Org.nr: {orgnr}</div>
          <div class="card-details">
            <div class="detail-row">📍 <span>{adresse}</span></div>
            <div class="detail-row">📞 <span>{telefon}</span></div>
            <div class="detail-row">🌐 <span>{nettside}</span></div>
          </div>
          <div class="services-row">
            {tjeneste_tags}
          </div>
          <div class="card-footer">
            <button class="btn-primary">Se detaljer</button>
            <button class="btn-secondary">Kart</button>
          </div>
        </div>"""


def oppdater_html(html_fil: Path, klinikkort_html: str, antall: int) -> None:
    innhold = html_fil.read_text(encoding="utf-8")

    # Erstatt klinikkgrid-innholdet
    import re
    nytt_grid = (
        f'<div class="clinics-grid">\n{klinikkort_html}\n      </div>'
    )
    innhold = re.sub(
        r'<div class="clinics-grid">.*?</div>(?=\s*</section>)',
        nytt_grid,
        innhold,
        flags=re.DOTALL,
    )

    # Oppdater antall klinikker i hero-statistikk
    innhold = re.sub(
        r'(<span class="stat-number">)\d+(</span>\s*<span class="stat-label">Registrerte klinikker)',
        rf'\g<1>{antall}\2',
        innhold,
    )

    html_fil.write_text(innhold, encoding="utf-8")
    print(f"  Oppdatert: {html_fil}")


def main():
    parser = argparse.ArgumentParser(description="Hent klinikker fra Brønnøysund")
    parser.add_argument("--kommune", help="Kommunenummer, f.eks. 3005")
    parser.add_argument("--kommunenavn", help="Kommunenavn, f.eks. drammen", default=None)
    parser.add_argument("--kun-data", action="store_true", help="Skriv kun JSON, ikke HTML")
    args = parser.parse_args()

    kommunenummer = args.kommune
    if not kommunenummer and args.kommunenavn:
        kommunenummer = KOMMUNER.get(args.kommunenavn.lower())
    if not kommunenummer:
        print(f"Kjente kommuner: {', '.join(KOMMUNER)}")
        print("Bruk --kommune <nr> eller --kommunenavn <navn>")
        return

    kommunenavn = args.kommunenavn or kommunenummer
    print(f"\nHenter klinikker for kommune {kommunenummer} ({kommunenavn.title()})...")

    alle = []
    for nace in NACE_KODER:
        alle.extend(hent_alle_for_nace(nace, kommunenummer))

    # Dedupliser på org.nr
    sett = {}
    for e in alle:
        sett[e["organisasjonsnummer"]] = e
    unike = list(sett.values())
    unike.sort(key=lambda e: e.get("navn", ""))

    print(f"\nTotalt {len(unike)} unike klinikker funnet.")

    # Lagre rådata
    rot = Path(__file__).parent.parent
    data_mappe = rot / "scripts" / "data"
    data_mappe.mkdir(exist_ok=True)
    json_fil = data_mappe / f"klinikker_{kommunenavn.lower()}.json"
    json_fil.write_text(json.dumps(unike, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"  Rådata lagret: {json_fil}")

    if args.kun_data:
        return

    # Generer HTML-klinikkort
    alle_kort = "\n".join(generer_klinikkort(e) for e in unike)

    # Finn og oppdater relevante HTML-filer
    for mappe in rot.iterdir():
        if not mappe.is_dir() or mappe.name.startswith("_") or mappe.name.startswith("."):
            continue
        if kommunenavn.lower() in mappe.name.lower():
            html_fil = mappe / "index.html"
            if html_fil.exists():
                oppdater_html(html_fil, alle_kort, len(unike))

    print("\nFerdig!")


if __name__ == "__main__":
    main()
